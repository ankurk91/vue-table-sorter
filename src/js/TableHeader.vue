<template>
  <thead>
    <tr>
      <slot/>
    </tr>
  </thead>
</template>

<script>
import store from './tableStore';

export default {
  name: 'TableHeader',
  props: {
    sortBy: {
      type: String,
      required: true,
    },
    sortOrder: {
      type: String,
      required: true,
      validator: value => ['asc', 'desc'].includes(value?.toLowerCase())
    }
  },
  data() {
    return {
      store: store
    }
  },
  provide() {
    return {
      store: this.store
    }
  },
  created() {
    this.store.sortBy = this.sortBy
    this.store.sortOrder = this.sortOrder
  },
  mounted() {
    this.$watch('store', this.onChange, {
      deep: true
    })
  },
  methods: {
    onChange(newValue) {
      this.$emit('sort', newValue)
    }
  },
}
</script>
