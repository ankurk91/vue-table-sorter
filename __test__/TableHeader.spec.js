import {mount} from '@vue/test-utils';
import TableHeader from '../src/js/TableHeader.vue';

describe('TableHeader component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(TableHeader, {
      propsData: {
        sortBy: 'name',
        sortOrder: 'desc',
      },
      slots: {
        default: '<th class="text-center">Actions</th>'
      }
    });
  });

  test('renders with props and slots', async () => {
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$el).toMatchSnapshot();
    expect(wrapper.find('th').exists()).toBe(true);
    wrapper.unmount();
  });

  test('emits event on store change', async () => {
    wrapper.vm.store.sortBy = 'email';
    wrapper.vm.store.sortOrder = 'desc';

    await wrapper.vm.$nextTick();

    expect(wrapper.emitted().sort).toBeTruthy();
    expect(wrapper.emitted().sort[0][0]).toEqual({
      sortBy: 'email',
      sortOrder: 'desc'
    });
  })

});
