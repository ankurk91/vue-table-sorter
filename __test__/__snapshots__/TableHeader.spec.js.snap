// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableHeader component renders with props and slots 1`] = `
<thead>
  <tr>
    <th
      class="text-center"
    >
      Actions
    </th>
  </tr>
</thead>
`;
